File: snail-race-bet/typescript/src/domain/RaceResultProvider.ts
export class Snail {
    constructor(
        public readonly number: number,
        public readonly name: string,
    ) {
    }
}

export class Podium {
    constructor(
        public readonly first: Snail,
        public readonly second: Snail,
        public readonly third: Snail
    ) {
    }

    static NineHeightSeven() {
        return new Podium(new Snail(9, 'Nine'), new Snail(8, 'Height'), new Snail(7, 'Seven'));
    }

    static FiveSixSeven() {
        return new Podium(new Snail(5, 'Speedy'), new Snail(6, 'Naruto'), new Snail(7, 'Michka'));
    }
}

export class SnailRace {
    constructor(
        public readonly raceId:number,
        public readonly timestamp:number,
        public readonly podium:Podium
    ) {
    }

}

function sortFromMoreToLessRecent(a: SnailRace, b:SnailRace) {
    return b.timestamp - a.timestamp
}

export class SnailRaces {
    public readonly races: SnailRace[];

    constructor(races:Array<SnailRace>) {
        this.races = new Array(...races).sort(sortFromMoreToLessRecent)
    }

    static withAdditionalResult(snailRaces: SnailRaces, raceId: number, datetime: number, podium: Podium) {
        return new SnailRaces([...snailRaces.races, new SnailRace(raceId, datetime, podium)]);
    }
}

export interface RaceResultProvider {
    races(): Promise<SnailRaces>
}

File: snail-race-bet/python/src/domain/race_result_provider.py
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import List


@dataclass
class Snail:
    number: int
    name: str


@dataclass
class Podium:
    first: Snail
    second: Snail
    third: Snail


@dataclass
class SnailRace:
    race_id: int
    timestamp: int
    podium: Podium


class SnailRaces:
    races: List[SnailRace]

    def __init__(self):
        self.races = []

    def with_additional_result(self, race_id: int, datetime: int, podium: Podium) -> 'SnailRaces':
        self.races.append(SnailRace(race_id, datetime, podium))
        return self

    def get_last_race(self) -> SnailRace:
        if not self.races:
            return None
        return self.races[0]

class RaceResultProvider(ABC):

    @abstractmethod
    def races(self) -> SnailRaces:
        pass

